src/tools/dashboard/src/reference.rs

```diff
@@ -0,0 +1,287 @@
+        ]
+        .iter()
+        .collect(),
+        [
+            "src",
+            "test",
+            "ref",
+            "Statements and expressions",
+            "Expressions",
+            "Method call expressions",
+            "10.rs",
+        ]
+        .iter()
+        .collect(),
+    ];
+
+    for test in loop_tests {
+        let code = fs::read_to_string(&test).unwrap();
+        let code = format!("// cbmc-flags: --unwind 1 --unwinding-assertions\n{}", code);
+        fs::write(&test, code).unwrap();
+    }
+}
+
+/// Runs `compiletest` on the `suite` and logs the results to `log_path`.
+fn run_examples(suite: &str, log_path: &Path) {
+    // Before executing this program, `cargo` populates the environment with
+    // build configs. `x.py` respects those configs, causing a recompilation
+    // of `rustc`. This is not a desired behavior, so we remove those configs.
+    let filtered_env: HashMap<String, String> = env::vars()
+        .filter(|&(ref k, _)| {
+            !(k.contains("CARGO") || k.contains("LD_LIBRARY_PATH") || k.contains("RUST"))
+        })
+        .collect();
+    let mut cmd = Command::new([".", "x.py"].iter().collect::<PathBuf>());
+    cmd.args([
+        "test",
+        suite,
+        "-i",
+        "--stage",
+        "1",
+        "--test-args",
+        "--logfile",
+        "--test-args",
+        log_path.to_str().unwrap(),
+    ]);
+    cmd.env_clear().envs(filtered_env);
+    cmd.stdout(Stdio::null());
+
+    cmd.spawn()
+        .expect(&format!("Error: failed to spawn process:\n{:?}", cmd))
+        .wait()
+        .expect(&format!("Error: spawned process did not run:\n{:?}", cmd));
+}
+
+/// Creates a new [`Tree`] from a path, `ns`, and a test result, `p`.
+fn tree_from_path(mut ns: Vec<String>, p: bool) -> dashboard::Tree {
+    assert!(ns.len() > 0, "Error: `ns` must contain at least 1 element.");
+    let mut tree = dashboard::Tree::new(
+        dashboard::Node::new(ns.pop().unwrap(), if p { 1 } else { 0 }, if p { 0 } else { 1 }),
+        vec![],
+    );
+    for _ in 0..ns.len() {
+        tree = dashboard::Tree {
+            data: dashboard::Node {
+                name: ns.pop().unwrap(),
+                num_pass: tree.data.num_pass,
+                num_fail: tree.data.num_fail,
+            },
+            children: vec![tree],
+        };
+    }
+    tree
+}
+
+/// Parses and generates a dashboard from the log output of `compiletest` in
+/// `path`.
+fn parse_log(path: &Path) -> dashboard::Tree {
+    let file = fs::File::open(path).expect(&format!("Error: failed to open log file: {:?}", path));
+    let reader = BufReader::new(file);
+    let mut tests = dashboard::Tree {
+        data: dashboard::Node { name: String::from("ref"), num_pass: 0, num_fail: 0 },
+        children: vec![],
+    };
+    for line in reader.lines() {
+        let (ns, l) = parse_log_line(&line.unwrap());
+        tests = dashboard::Tree::merge(tests, tree_from_path(ns, l)).unwrap();
```

Comment:
Addressed above (when do you ever return `None`), but if you are internally using options, you should probably be handling them with better errors here.

Reply:
That's not necessary. All generated test trees have `ref` (the new test suite) as their root.